docs/decisions/0016-event-design-practices.rst

```diff
@@ -0,0 +1,127 @@
+
+Event Purpose and Content
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- An event should describe as accurately as possible what happened (what) and why it happened (why). It must contain enough information for consumers to understand the message. For instance, if an event is about a user enrollment, it should contain the user's data, the course data, and the enrollment status and the event should be named accordingly.
+- Avoid immediately contacting the source service to retrieve additional information. Instead, consider adding the necessary information to the event payload by managing the granularity of the event. If the event requires additional information, consider adding a field to the event that contains the necessary information. This will reduce the number of dependencies between services and make the event more self-contained.
+- Keep the event size small. Avoid adding unnecessary information to the event. If the information is not necessary for consumers to react to the event, consider removing it.
+- Avoid adding flow-control information or business logic to events. Events should be solely a representation of what took place. If a field is necessary to control the behavior of the consumer, consider moving it to the consumer side. If adding additional data to the event is absolutely necessary document the reasoning behind it and carefully study the use case and implications.
+
+Here is an example of an event that follows these practices which is emitted when the a user registers:
+
+.. code-block:: python
+
+    # Location openedx_events/learning/signal.py
+    # .. event_type: org.openedx.learning.student.registration.completed.v1
+    # .. event_name: STUDENT_REGISTRATION_COMPLETED
+    # .. event_description: emitted when the user registration process in the LMS is completed.
+    # .. event_data: UserData
+    STUDENT_REGISTRATION_COMPLETED = OpenEdxPublicSignal(
+        event_type="org.openedx.learning.student.registration.completed.v1",
+        data={
+            "user": UserData,
+        }
+    )
+
+Where:
+
+- The event name indicates what happened: ``STUDENT_REGISTRATION_COMPLETED``.
+- The event description explains why the event happened: ``emitted when the user registration process in the LMS is completed``.
+- The event data contains data directly related to what happened ``UserData`` which should contain the necessary information to understand the event, like the username and email of the user.
+
+Responsibility and Granularity
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Design events with a single responsibility in mind. Each event should represent a single action or fact that happened in the system. If an event contains multiple actions, consider splitting it into multiple events. For instance, if the course grade is updated to pass or fail, there should be two events: one for the pass action and another for the fail action.
```

Comment:
I know this ADR is about the event design itself, but somewhere you may want to consider adding a note like the following:

Events that are split across multiple actions is an exceptional case where the same event queue/topic should be used to help maintain order across these events.

Reply:
This note's pretty useful! Thanks. I'll make sure to add it right away.